docs/adr/0013-data-serialization-within-on-chain-validators.md

```diff
@@ -0,0 +1,60 @@
+# 13. Data Serialization Within On-Chain Validators
+
+Date: 2021-12-21
+
+## Status
+
+Proposed
+
+## Context
+
+In Hydra, during the `Close` and `Contest` transitions, one must verify, within
+on-chain validators, that a certain piece of data has been multi-signed by all
+head participants. While verifying a multi-signature performed via [MuSig2][]
+(which can be made Schnorr-compatible) is relatively easy and can rely on
+existing Plutus built-ins; producing the payload / pre-image that was signed is
+problematic for there's no Plutus built-ins regarding data serialization. 
+
+Incidentally, event though there exists quite simple and compact
+(implementation-wise) serialization algorithms (e.g. CBOR), this is path we do
+not want to follow as there's a high chance to increase the validator size far
+above an acceptable limit. 
+
+Hence, how to obtain arbitrary serialized data within an on-chain validator?
+
+## Decision
+
+#### Overview
+
+In Cardano, transactions may carry information in various ways and in particular, one must provide Plutus data as part of a transaction witness set. Those data are made available to the underlying validator script context as a (key, value) list where keys are data hashes and value the data. It's important to note that the correspondence between a hash and its data is 
+verified by the ledger during phase-1 validations; 
+
+We want to leverage this data lookup table to pass arbitrary data and their corresponding hashes to a validator. This effectively means that we introduce an extra indirection in the redeemer of the `Close` and `Context` transition. Indeed, instead of passing the full data as the redeemer, we can only give a hash which can be looked up from the script context to obtain its corresponding data. This can be achieved with the following on-chain function:
```

Comment:
I don't know what are the requirements of the `Close` and `Context` transactions so the question might be stupid... But if the multi-signed data piece isn't duplicated, is representing it as a datum more efficient than a redeemer? Both require the data in the transaction witness set but finding data in the script context is costly inside validator scripts. Redeemers also don't require storing unused data (datum hash of the multi-signed data piece) in the next UTxO set.

Reply:
I am not quite sure what you mean by "isn't duplicated"?

The multisigned payloads are produced off-chain, as part of the running head. It is basically what entitles participants to move the contract on-chain. Thus, it is precisely something we cannot store on-chain upfront of course. Nor it is something we can pass as-is for redeemer to the script (because of the surrounding information that come with the multisigned payload).

I would need to double check that but, as far as I remember, the ledger rules also forbid to add supplementary 'unused' redeemer to the script data.

Reply:
> I am not quite sure what you mean by "isn't duplicated"?

Sorry for the vague description. By that, I meant redeemers are stored per input (more precisely per pointer as things like minting policies have redeemers too) while datums are stored per unique hash. If the multi-signed data is only used in the validation of one input then the storage in the witness set is roughly the same. Datums are only "lighter" when that same data is used in the validation of multiple inputs. In this context, I guess that the multi-signed data is only used in the validation of its specific head UTxO, hence no storage gains from the datum way.

> Nor it is something we can pass as-is for redeemer to the script (because of the surrounding information that come with the multisigned payload).

I'm still confused why you can represent something as a datum but not as a redeemer. Aren't they represented in the same `Data` type at both the ledger and script levels? 

> I would need to double check that but, as far as I remember, the ledger rules also forbid to add supplementary 'unused' redeemer to the script data. Hence the use of a datum for that.

Isn't the redeemer used in the validation script for the multi-signature check?